service/topology_coordinator.cc

```diff
@@ -1117,7 +1117,12 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
                                 excluded_new++;
                             }
                         }
-                        if (excluded_new > excluded_old) {
+                        // Cannot revert if this is intra-node migration.
+                        // We will lose data if the migrating node restarted, in which case the leaving replica
+                        // doesn't contain writes anymore as sstables are attached only on the pending replica.
```

Comment:
Could we not attach sstables on both replicas?

Reply:
I wonder what happen if migrating node is restarted without revert after streaming stage. we'll attach both sstables of both pending and replica into each of them, which means duplication (not correctness issue), but that should be rare enough, as intra migration should complete quickly. Also, we need to verify what happens in sstable_directory, if sharder says sstables belong to two shards (pending and leaving), it may want to trigger sstable resharding on them. for tablets, we can treat it differently. Or sharder (used by dht::ring_position_range_vector_sharder) will point only to pending? In that case, we could also lose data if migratin node restarts before streaming stage, right?

Reply:
I thought we were attaching sstables to both replicas, but apparently we use shard_for_reads() that return only one shard based on migration stage. SSTable has its owner shard(s) computed using ring_position_range_vector_sharder.

But now I think I understand why it's not needed. After streaming stage, all new reads should only be targeting the new replica. Could it happen that somehow old reads somehow reach the leaving replica in write_both_read_new stage (after restart)? If I understand correctly, it won't contain all writes as sstables were only loaded to pending replica on restart.

Reply:
I think it's not a problem as requests will be correctly forwarded to pending replica in the replica side.

Reply:
> But now I think I understand why it's not needed. After streaming stage, all new reads should only be targeting the new replica. Could it happen that somehow old reads somehow reach the leaving replica in write_both_read_new stage (after restart)? If I understand correctly, it won't contain all writes as sstables were only loaded to pending replica on restart.

If routing to shards happens only from the local node based on current erm, then all reads should use the new shard.